### ISSUE TITLE
[BUG] Differences between old_log_prob and log_prob in grpo on-policy setting and fsdp backend for multimodal training like qwen-vl with geo3k(serious problem).

### ISSUE THREAD

**Issue by Serge-weihao** [verl-project/verl#2996]
I set train_batch_size==ppo_mini_batch_size in grpo on-policy setting but find the differences between old_log_prob and log_prob as below. This bug occurs in all versions after v0.4.0.
 
verl/verl/workers/actor/dp_actor.py diff_list [[tensor(-0.1407, device='cuda:0'), tensor(-0.1407, device='cuda:0', grad_fn=<MeanBackward0>)], **[tensor(-0.1376, device='cuda:0'), tensor(-0.1372, device='cuda:0', grad_fn=<MeanBackward0>)],** [tensor(-0.1332, device='cuda:0'), tensor(-0.1331, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1780, device='cuda:0'), tensor(-0.1778, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1602, device='cuda:0'), tensor(-0.1603, device='cuda:0', grad_fn=<MeanBackward0>)], **[tensor(-0.2336, device='cuda:0'), tensor(-0.2338, device='cuda:0', grad_fn=<MeanBackward0>)]**, [tensor(-0.2920, device='cuda:0'), tensor(-0.2924, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3069, device='cuda:0'), tensor(-0.3065, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2686, device='cuda:0'), tensor(-0.2686, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3267, device='cuda:0'), tensor(-0.3269, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1035, device='cuda:0'), tensor(-0.1032, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1377, device='cuda:0'), tensor(-0.1377, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1768, device='cuda:0'), tensor(-0.1770, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3130, device='cuda:0'), tensor(-0.3134, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1737, device='cuda:0'), tensor(-0.1738, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.0879, device='cuda:0'), tensor(-0.0881, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2116, device='cuda:0'), tensor(-0.2115, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1952, device='cuda:0'), tensor(-0.1955, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1552, device='cuda:0'), tensor(-0.1552, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1190, device='cuda:0'), tensor(-0.1189, device='cuda:0', grad_fn=<MeanBackward0>)], **[tensor(-0.2715, device='cuda:0'), tensor(-0.2720, device='cuda:0', grad_fn=<MeanBackward0>)]**, [tensor(-0.2151, device='cuda:0'), tensor(-0.2154, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.0655, device='cuda:0'), tensor(-0.0653, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1575, device='cuda:0'), tensor(-0.1575, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.0805, device='cuda:0'), tensor(-0.0801, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1391, device='cuda:0'), tensor(-0.1392, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1818, device='cuda:0'), tensor(-0.1813, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.4254, device='cuda:0'), tensor(-0.4254, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3124, device='cuda:0'), tensor(-0.3122, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1048, device='cuda:0'), tensor(-0.1047, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1301, device='cuda:0'), tensor(-0.1301, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1658, device='cuda:0'), tensor(-0.1655, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1548, device='cuda:0'), tensor(-0.1547, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3397, device='cuda:0'), tensor(-0.3401, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1230, device='cuda:0'), tensor(-0.1231, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1640, device='cuda:0'), tensor(-0.1646, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2249, device='cuda:0'), tensor(-0.2251, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3790, device='cuda:0'), tensor(-0.3792, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2259, device='cuda:0'), tensor(-0.2258, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.4029, device='cuda:0'), tensor(-0.4029, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.0993, device='cuda:0'), tensor(-0.0990, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1822, device='cuda:0'), tensor(-0.1822, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1235, device='cuda:0'), tensor(-0.1234, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.5672, device='cuda:0'), tensor(-0.5668, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1263, device='cuda:0'), tensor(-0.1261, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2395, device='cuda:0'), tensor(-0.2395, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2349, device='cuda:0'), tensor(-0.2349, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.4052, device='cuda:0'), tensor(-0.4048, device='cuda:0',grad_fn=<MeanBackward0>)], [tensor(-0.3712, device='cuda:0'), tensor(-0.3711, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2008, device='cuda:0'), tensor(-0.2010, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2720, device='cuda:0'), tensor(-0.2719, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.0967, device='cuda:0'), tensor(-0.0964, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2574, device='cuda:0'), tensor(-0.2574, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3928, device='cuda:0'), tensor(-0.3927, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1153, device='cuda:0'), tensor(-0.1153, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1113, device='cuda:0'), tensor(-0.1115, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2600, device='cuda:0'), tensor(-0.2603, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.4577, device='cuda:0'), tensor(-0.4571, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1976, device='cuda:0'), tensor(-0.1979, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1522, device='cuda:0'), tensor(-0.1523, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1911, device='cuda:0'), tensor(-0.1911, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1475, device='cuda:0'), tensor(-0.1478, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.7045, device='cuda:0'), tensor(-0.7042, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1842, device='cuda:0'), tensor(-0.1842, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3147, device='cuda:0'), tensor(-0.3149, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.6061, device='cuda:0'), tensor(-0.6060, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2166, device='cuda:0'), tensor(-0.2166, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2871, device='cuda:0'), tensor(-0.2866, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3649, device='cuda:0'), tensor(-0.3654, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2259, device='cuda:0'), tensor(-0.2260, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1422, device='cuda:0'), tensor(-0.1422, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1319, device='cuda:0'), tensor(-0.1320, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1774, device='cuda:0'), tensor(-0.1767, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1478, device='cuda:0'), tensor(-0.1478, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2243, device='cuda:0'), tensor(-0.2240, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1497, device='cuda:0'), tensor(-0.1500, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1198, device='cuda:0'), tensor(-0.1198, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.2363, device='cuda:0'), tensor(-0.2368, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.1806, device='cuda:0'), tensor(-0.1808, device='cuda:0', grad_fn=<MeanBackward0>)], [tensor(-0.3737, device='cuda:0'), tensor(-0.3737, device='cuda:0', grad_fn=<MeanBackward0>)]]

**Comment by Serge-weihao** [comment]
this bug only happens with multimodal inputs, if we train qwen-vl with gsm8k, it will not occur.

**Comment by wonderseen** [comment]
Curious if increasing tp or sp would enlarge the log_prob difference?

**Comment by Serge-weihao** [comment]
> Curious if increasing tp or sp would enlarge the log_prob difference?






> Curious if increasing tp or sp would enlarge the log_prob difference?

As far as I can recall, there is no significant difference regarding sp==1 after sp==2 in the dense model like qwen-vl, but I haven't had access to devices to re-verify it recently. In addition, the diff will increase with larger model size and the diff of moe-vl model is significantly larger than dense-vl model.

**Comment by Kirrito-k423** [comment]
中文：感谢报告。这个问题不建议直接关闭：从当前公开证据看，`old_log_prob` / `log_prob` 在 on-policy 设置下不一致仍属于未完全闭环的 logprob 一致性问题。相关讨论仍然 open，例如 #6280 报告了 micro-batch 不一致导致的 on-policy mismatch，并给出临时 workaround：让 `actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu` 与 `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu` 保持一致；#6240 也讨论了 rollout `processed_logprobs` 与 trainer forward logprobs 的语义差异。vLLM 侧也有 RFC vllm-project/vllm#42259 在跟踪 logprobs/logits 语义与确定性问题。

为了进一步判断 #2996 是否与这些已知问题同根，能否补充以下信息：1) 完整 verl commit/version、vLLM version、transformers version；2) 完整训练配置，特别是 `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu`、`actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu`、`actor_rollout_ref.rollout.logprobs_mode`、`actor_rollout_ref.model.use_remove_padding`、TP/SP/DP 设置；3) 是否在将 rollout log_prob micro-batch size 设为与 actor ppo micro-batch size 相同后仍能复现；4) 一个最小可运行脚本或配置片段，以及对应的 `actor/pg_clipfrac`、`actor/ppo_kl`、rollout/old/new logprob diff 日志。

English: Thanks for the report. I would not close this yet: the public evidence suggests this is still part of an unresolved logprob-consistency class of issues. Related open discussions include #6280, which reports an on-policy mismatch caused by rollout/actor micro-batch differences and suggests the immediate workaround of setting `actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu` equal to `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu`; #6240 also discusses the semantic mismatch between rollout `processed_logprobs` and trainer-side forward logprobs. On the vLLM side, vllm-project/vllm#42259 is tracking logprob/logit semantics and determinism across the ecosystem.

To determine whether #2996 has the same root cause, could you please provide: 1) exact verl commit/version, vLLM version, and transformers version; 2) the full training config, especially `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu`, `actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu`, `actor_rollout_ref.rollout.logprobs_mode`, `actor_rollout_ref.model.use_remove_padding`, and TP/SP/DP settings; 3) whether the mismatch still reproduces after setting the rollout log_prob micro-batch size equal to the actor PPO micro-batch size; 4) a minimal runnable script/config plus the corresponding `actor/pg_clipfrac`, `actor/ppo_kl`, and rollout/old/new logprob diff logs.
This response was generated by a scheduled Codex GPT5.5 issue processing task.
